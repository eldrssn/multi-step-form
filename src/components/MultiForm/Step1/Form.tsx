import { FC } from 'react';
import { Formik, FormikProps } from 'formik';

import { TextInput } from '@/components/ui/TextInput';
import { Button } from '@/components/ui/Button';

import { TStep, TStep1 } from '../types';
import { TErrors } from './types';
import { step1 } from './constants';
import styles from './Form.module.scss';

const validate = (values: TStep1) => {
  const errors: TErrors = {};

  if (!values.name) {
    errors.name = 'This field is required';
  }

  if (!values.email) {
    errors.email = 'This field is required';
  } else if (!/^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,4}$/i.test(values.email)) {
    errors.email = 'Invalid email address';
  }

  if (!values.phone) {
    errors.phone = 'This field is required';
  }

  return errors;
};

const Form: FC<TStep> = ({ handleNext, data: { step1Data } }) => (
  <Formik
    initialValues={
      step1Data || {
        name: '',
        email: '',
        phone: '',
      }
    }
    validate={validate}
    onSubmit={(values) => {
      handleNext && handleNext({ step1Data: { ...values } });
    }}
  >
    {(props: FormikProps<TStep1>) => (
      <form onSubmit={props.handleSubmit} className={styles.form}>
        {step1.map((input) => (
          <TextInput input={input} key={input.name} />
        ))}
        <div className={styles.buttons}>
          <Button type="submit" action="next" />
        </div>
      </form>
    )}
  </Formik>
);

export { Form };
