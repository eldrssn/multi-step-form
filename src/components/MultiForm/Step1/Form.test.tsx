import { render, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { Form } from './Form';

describe('Form', () => {
  test('renders the form with TextInput and Button', () => {
    const { getByPlaceholderText, getByText, container } = render(
      <Form
        handleNext={jest.fn()}
        data={{ step1Data: null, step2Data: null, step3Data: null }}
      />
    );

    const name = getByPlaceholderText(/e.g. stephen king/i);
    const email = getByPlaceholderText(/e.g. stephenking@lorem.com/i);
    const phone = getByPlaceholderText('e.g. +1 234 567 890');
    const buttonElement = getByText(/next/i);

    expect(name).toBeInTheDocument();
    expect(email).toBeInTheDocument();
    expect(phone).toBeInTheDocument();
    expect(buttonElement).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  test('submits the form with valid data', async () => {
    const handleNextMock = jest.fn();
    const { getByPlaceholderText, getByText } = render(
      <Form
        handleNext={handleNextMock}
        data={{ step1Data: null, step2Data: null, step3Data: null }}
      />
    );

    const user = userEvent.setup();

    await user.type(getByPlaceholderText(/e.g. stephen king/i), 'John Doe');
    await user.type(
      getByPlaceholderText(/e.g. stephenking@lorem.com/i),
      'john.doe@example.com'
    );
    await user.type(getByPlaceholderText('e.g. +1 234 567 890'), '1234567890');

    await user.click(getByText(/next/i));

    await waitFor(() => {
      expect(handleNextMock).toHaveBeenCalledWith({
        step1Data: {
          name: 'John Doe',
          email: 'john.doe@example.com',
          phone: 1234567890,
        },
      });
    });
  });

  test('submits the form with no name', async () => {
    const handleNextMock = jest.fn();
    const { getByPlaceholderText, getByText, findByText } = render(
      <Form
        handleNext={handleNextMock}
        data={{ step1Data: null, step2Data: null, step3Data: null }}
      />
    );

    const user = userEvent.setup();

    await user.type(
      getByPlaceholderText(/e.g. stephenking@lorem.com/i),
      'john.doe@example.com'
    );
    await user.type(getByPlaceholderText('e.g. +1 234 567 890'), '1234567890');

    await user.click(getByText(/next/i));

    await waitFor(async () => {
      expect(await findByText(/This field is required/i)).toBeInTheDocument();
      expect(handleNextMock).not.toBeCalled();
    });
  });

  test('submits the form with no valid email', async () => {
    const handleNextMock = jest.fn();
    const { getByPlaceholderText, getByText, findByText } = render(
      <Form
        handleNext={handleNextMock}
        data={{ step1Data: null, step2Data: null, step3Data: null }}
      />
    );

    const user = userEvent.setup();

    await user.type(getByPlaceholderText(/e.g. stephen king/i), 'John Doe');
    await user.type(
      getByPlaceholderText(/e.g. stephenking@lorem.com/i),
      'john.doe@example'
    );
    await user.type(getByPlaceholderText('e.g. +1 234 567 890'), '1234567890');

    await user.click(getByText(/next/i));

    await waitFor(async () => {
      expect(await findByText(/Invalid email address/i)).toBeInTheDocument();
      expect(handleNextMock).not.toBeCalled();
    });
  });

  test('submits the form with existing data', async () => {
    const { getByPlaceholderText } = render(
      <Form
        handleNext={jest.fn()}
        data={{
          step1Data: {
            name: 'John Doe',
            email: 'john.doe@example.com',
            phone: '1234567890',
          },
          step2Data: null,
          step3Data: null,
        }}
      />
    );

    expect(getByPlaceholderText(/e.g. stephen king/i)).toHaveValue('John Doe');
    expect(getByPlaceholderText(/e.g. stephenking@lorem.com/i)).toHaveValue(
      'john.doe@example.com'
    );
    expect(getByPlaceholderText('e.g. +1 234 567 890')).toHaveValue(1234567890);
  });
});
