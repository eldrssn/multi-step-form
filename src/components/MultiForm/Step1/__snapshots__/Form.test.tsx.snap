// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form renders the form with TextInput and Button 1`] = `
<div>
  <form
    class="form"
  >
    <div
      class="input"
    >
      <div
        class="label_box"
      >
        <label
          for="name"
        >
          Name
        </label>
      </div>
      <input
        id="name"
        label="Name"
        name="name"
        placeholder="e.g. Stephen King"
        type="text"
        value=""
      />
    </div>
    <div
      class="input"
    >
      <div
        class="label_box"
      >
        <label
          for="email"
        >
          Email Address
        </label>
      </div>
      <input
        id="email"
        label="Email Address"
        name="email"
        placeholder="e.g. stephenking@lorem.com"
        type="text"
        value=""
      />
    </div>
    <div
      class="input"
    >
      <div
        class="label_box"
      >
        <label
          for="phone"
        >
          Phone Number
        </label>
      </div>
      <input
        id="phone"
        label="Phone Number"
        name="phone"
        placeholder="e.g. +1 234 567 890"
        type="number"
        value=""
      />
    </div>
    <div
      class="buttons"
    >
      <button
        class="button button_next"
        type="submit"
      >
        Next Step
      </button>
    </div>
  </form>
</div>
`;
